test/ash/data_layer/mnesia_test.exs

```diff
@@ -0,0 +1,176 @@
+
+    mnesia do
+      table :test_users
+    end
+
+    actions do
+      defaults [:read, :destroy, create: :*]
+
+      update :update do
+        accept :*
+        require_atomic? false
+      end
+    end
+
+    identities do
+      identity :unique_name, [:name], pre_check_with: Domain
+    end
+
+    attributes do
+      uuid_primary_key :id, writable?: true
+      attribute :name, :string, public?: true
+      attribute :age, :integer, public?: true
+      attribute :title, :string, public?: true
+      attribute :roles, {:array, :atom}, public?: true
+    end
+  end
+
+  describe "create/2" do
+    test "it creates a user" do
+      user = %{name: "John", age: 30, title: "Developer", roles: [:admin, :user]}
+      assert {:ok, created_user} = Ash.create(MnesiaTestUser, user)
+
+      assert %MnesiaTestUser{
+               id: _,
+               name: "John",
+               age: 30,
+               title: "Developer",
+               roles: [:admin, :user]
+             } = created_user
+    end
+  end
+
+  describe "bulk_create/3" do
+    test "can? bulk_create" do
+      assert MnesiaDataLayer.can?(:test_resource, :bulk_create)
+    end
+
+    test "bulk_create/3 without returning records" do
+      resp =
+        [
+          %{name: "John", age: 30, title: "Developer", roles: [:admin, :user]},
+          %{name: "Jane", age: 25, title: "Designer", roles: [:user]},
+          %{name: "Bob", age: 35, title: "Manager", roles: [:admin, :manager]}
+        ]
+        |> Ash.bulk_create(MnesiaTestUser, :create)
+
+      assert %Ash.BulkResult{
+               status: :success,
+               errors: [],
+               records: nil,
+               notifications: nil,
+               error_count: 0
+             } = resp
+    end
+
+    test "bulk_create/3 WITH returning records" do
+      resp =
+        [
+          %{name: "John", age: 30, title: "Developer", roles: [:admin, :user]},
+          %{name: "Jane", age: 25, title: "Designer", roles: [:user]},
+          %{name: "Bob", age: 35, title: "Manager", roles: [:admin, :manager]}
+        ]
+        |> Ash.bulk_create(MnesiaTestUser, :create, return_records?: true)
+
+      assert %Ash.BulkResult{
+               status: :success,
+               errors: [],
+               records: [
+                 %MnesiaTestUser{
+                   name: "John",
+                   age: 30,
+                   title: "Developer",
+                   roles: [:admin, :user]
+                 },
+                 %MnesiaTestUser{name: "Jane", age: 25, title: "Designer", roles: [:user]},
+                 %MnesiaTestUser{
+                   name: "Bob",
+                   age: 35,
+                   title: "Manager",
+                   roles: [:admin, :manager]
+                 }
+               ],
+               notifications: nil,
+               error_count: 0
+             } = resp
+    end
+
+    test "bulk_create/3 with UPSERT without returning records" do
+      user_a =
```

Comment:
This fixes the crux of the issue I was having on upsert by including the `:id` in the attributes.